### PR TITLE
[BGP Memory Leak] Add FRR memory checker in the continuous link flap test

### PR DESCRIPTION
**Pull request by congh-nvidia**
### Description of PR

Summary:
This is to cover the BGP memory leak test gap.
Previously we have an issue in FRR(https://github.com/FRRouting/frr/issues/15459), which caused memory leak.

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [x] 202405

#### What is the motivation for this PR?
This is to cover the BGP memory leak test gap.
#### How did you do it?
I added a checker for the FRR daemons(bgpd and zebra) used memory in the test test_cont_link_flap.py. The reason why I added the checker in test_cont_link_flap.py but not the stress route test, is because we already have some memory and cpu checkers in the link flap test. And the link flap also causes routes announce/withdraw like what we have in the stress route test.
#### How did you verify/test it?
Run the test and it passed.

### CODE REVIEW

**Comment by congh-nvidia** [comment]
Hi @saiarcot895 @lguohan @lipxu , as @dgsudharsan suggested, I added a bgp memory checker in the continuous link flap test to cover the issues like https://github.com/FRRouting/frr/issues/15459. 
But I can't reproduce the memory leak issue on public or our internal images because we don't have the configuration which caused the memory leak.
I tested this change on our testbed and the test passed, could you please review if it could help catch the memory leak on your side? Thanks.

File: tests/platform_tests/link_flap/test_cont_link_flap.py [hunk 1]
```diff
@@ -128,6 +139,18 @@ def test_cont_link_flap(self, request, duthosts, nbrhosts, enum_rand_one_per_hws
...
+        # Record BGP memory status at end
+        end_time_bgp_memory = self.get_bgp_memory_usage(duthost)
+        logging.info("BGP Memory Usage at end: \n%s", end_time_bgp_memory)
+
+        # Calculate diff in BGP memory
+        incr_bgp_memory = int(end_time_bgp_memory) - int(start_time_bgp_memory)
+        logging.info("BGP absolute difference: %d", incr_bgp_memory)
+
+        # Check BGP memory is not increased
+        pytest_assert(incr_bgp_memory <= 0,
```
Comment (lipxu) [inline tests/platform_tests/link_flap/test_cont_link_flap.py:151]: Is this check condition too strict? should we add a threshold to allow for a reasonable range of the increase?

File: tests/platform_tests/link_flap/test_cont_link_flap.py [hunk 2]
```diff
@@ -28,6 +28,13 @@ class TestContLinkFlap(object):
     TestContLinkFlap class for continuous link flap
     """
 
+    def get_bgp_memory_usage(self, duthost):
+        bgp_memory_output = duthost.shell('vtysh -c "show memory bgp"')["stdout"]
```
Comment (lipxu) [inline tests/platform_tests/link_flap/test_cont_link_flap.py:32]: Better to check more memory usage in FRR, like zebra

**Comment by lipxu** [comment]
Is it possible to make this function as a fixture and auto check the memory for all the test cases?

**Comment by congh-nvidia** [comment]
> Is it possible to make this function as a fixture and auto check the memory for all the test cases?

Hi @lipxu, I don't think it's a good approach, because:
1. If the checker fails due to some general bgp memory leak issue, all functional tests will be blocked.
2. In some tests, the bgp memory usage could be expected to increase.

File: tests/platform_tests/link_flap/test_cont_link_flap.py (hunk 1 above)
Comment (congh-nvidia) [reply tests/platform_tests/link_flap/test_cont_link_flap.py]: Done. Added a threshold same as the generic memory check.

File: tests/platform_tests/link_flap/test_cont_link_flap.py (hunk 2 above)
Comment (congh-nvidia) [reply tests/platform_tests/link_flap/test_cont_link_flap.py]: Done, added zebra.

**Comment by wsycqyz** [comment]
Have triggered to rerun the failed jobs.

**Comment by congh-nvidia** [comment]
Hi @wangxin, the checker failed due to "##[error]The job running on agent sonic-ubu000ROE ran longer than the maximum time of 240 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134". Could you please check? The failure is also observed in other PRs.

**Comment by bingwang-ms** [comment]
@lipxu Can you please review the change and approve? Thanks

File: tests/platform_tests/link_flap/test_cont_link_flap.py [hunk 3]
```diff
@@ -124,10 +139,34 @@ def test_cont_link_flap(self, request, duthosts, nbrhosts, enum_rand_one_per_hws
...
+            # Calculate diff in FRR daemon memory
+            incr_frr_daemon_memory = \
+                float(end_time_frr_daemon_memory[daemon]) - float(start_time_frr_daemon_memory[daemon])
+            logging.info(f"{daemon} absolute difference: %d", incr_frr_daemon_memory)
+
+            # Check FRR daemon memory only if it is increased else default to pass
+            if incr_frr_daemon_memory > 0:
+                percent_incr_frr_daemon_memory = \
+                    (incr_frr_daemon_memory / float(start_time_frr_daemon_memory[daemon])) * 100
+                logging.info(f"{daemon} memory percentage increase: %d", percent_incr_frr_daemon_memory)
```
Comment (lipxu) [inline tests/platform_tests/link_flap/test_cont_link_flap.py:166]: May you please share the typical value for this, thanks

**Review by lipxu — APPROVED** [review]
LGME

File: tests/platform_tests/link_flap/test_cont_link_flap.py (hunk 3 above)
Comment (congh-nvidia) [reply tests/platform_tests/link_flap/test_cont_link_flap.py:166]: > May you please share the typical value for this, thanks

This is the memory usage of a 4700 platform with t1-56-lag topology:
before link flap:
```
INFO     root:test_cont_link_flap.py:34 zebra memory status: 
System allocator statistics:
  Total heap allocated:  19 MiB
  Holding block headers: 225 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  12 MiB
  Free small blocks:     2320 bytes
  Free ordinary blocks:  7222 KiB
  Ordinary blocks:       15757
  Small blocks:          69
  Holding blocks:        29

INFO     root:test_cont_link_flap.py:34 bgpd memory status: 
System allocator statistics:
  Total heap allocated:  59 MiB
  Holding block headers: 165 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  56 MiB
  Free small blocks:     2608 bytes
  Free ordinary blocks:  2969 KiB
  Ordinary blocks:       5828
  Small blocks:          61
  Holding blocks:        115
```
after link flap:
```
INFO     root:test_cont_link_flap.py:34 zebra memory status: 
System allocator statistics:
  Total heap allocated:  22 MiB
  Holding block headers: 225 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  12 MiB
  Free small blocks:     2640 bytes
  Free ordinary blocks:  10297 KiB
  Ordinary blocks:       15823
  Small blocks:          72
  Holding blocks:        29

INFO     root:test_cont_link_flap.py:34 bgpd memory status: 
System allocator statistics:
  Total heap allocated:  60 MiB
  Holding block headers: 164 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  57 MiB
  Free small blocks:     2000 bytes
  Free ordinary blocks:  3018 KiB
  Ordinary blocks:       2740
  Small blocks:          56
  Holding blocks:        114
```

**Review by bingwang-ms — APPROVED** [review]

**Pull request merged**